README.md
# flutter_msg_engine

Flutter Message Engine. Register a message where you need it and send it anywhere in the future. It can be responded to and processed. Help application decoupling.

## Getting Started

This project is a starting point for a Dart
[package](https://flutter.io/developing-packages/),
a library module containing code that can be shared easily across
multiple Flutter or Dart projects.

For help getting started with Flutter, view our 
[online documentation](https://flutter.io/docs), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.


CHANGELOG.md
## 1.0.0

Introduction to optimization and modification.

## 0.0.3

Bug repair.

## 0.0.2

Add registers.

## 0.0.1

* TODO: Describe initial release.


test/flutter_msg_engine_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_msg_engine/flutter_msg_engine.dart';

void main() {
  int sMsgId = 999;
  int sMsgRef = 0;

  test('adds one to input values', () {
    MsgEngine.instance.regMsg(111, (MsgPack msg) {
      print(msg.toString());
    });
    MsgEngine.instance.regMsg(115, (MsgPack msg) {
      print("115: ${msg.toString()}");
    });

    MsgEngine.instance.regMsg(118, (MsgPack msg) {
      print("118: ${msg.data}");
    });

    print(MsgEngine().toString());
    MsgPack msg = const MsgBase(msgId: 123, sender: null);
    print("${msg.toString()}");

    MsgEngine.instance.sendMsgSync(const MsgTextBase(msgId: 111, text: "Hello MsgEngine."));
    MsgEngine.instance.sendMsg(const MsgTextBase(msgId: 112, text: "empty"));
    MsgEngine.instance.sendMsg(const MsgTextBase(msgId: 111, text: "Hello MsgEngine 2."));
    MsgEngine.instance.sendMsg(const MsgTextBase(msgId: 113, text: "..."));
    MsgEngine.instance.sendMsg(const MsgTextBase(msgId: 115, text: "Hello MsgEngine 3."));
    MsgEngine.instance.sendMsg(const MsgTextBase(msgId: 150, text: "message 150."));

    MsgEngine.instance.sendMsg(const MsgPackData<int>(msgId: 150, data: 150));
    MsgEngine.instance.sendMsg(const MsgPackData<int>(msgId: 118, data: 520));

    MsgEngine.instance.sendMsg(MsgPackData<String>(msgId: sMsgId - 1, data: "message id is $sMsgRef"));
    MsgEngine.instance.sendMsg(MsgPackData<String>(msgId: sMsgId, data: "message id is $sMsgRef"));

    sMsgRef++;
  });
}


lib/flutter_msg_engine.dart
import 'dart:async';
import 'dart:collection';

typedef MsgProcHandlerCallBack<T>(MsgPack<T> msg);

/// 消息包接口
abstract class MsgPack<T> {
  const MsgPack();

  /// 消息ID
  int get msgId;

  /// 消息发送者
  get sender;

  String toString();

  T get data;
}

/// 消息处理器接口
abstract class MsgProcHandler<T> {
  /// 处理消息
  void processMsg(MsgPack<T> msg);
}

/// 消息处理器接口
abstract class MsgHandler extends MsgProcHandler {
  /// 已经注册过的消息ID
  List<int> get msgIds;

  /// 发送消息
  void sendMsg(MsgPack msg);
}

/// 消息包抽象类
abstract class MsgPackBase<T> implements MsgPack<T> {
  /// 消息ID
  final int msgId;

  /// 消息发送者
  final sender;

  const MsgPackBase({this.msgId, this.sender});

  String toString() {
    return "msgId: $msgId, sender: $sender";
  }
}

/// 消息包基类
class MsgBase<T> extends MsgPackBase<T> {
  const MsgBase({int msgId, sender}) : super(msgId: msgId, sender: sender);

  T get data => getData();

  T getData() {
    return null;
  }
}

/// 文本消息基础类
class MsgTextBase extends MsgBase<String> {
  const MsgTextBase({this.text, int msgId, sender})
      : super(msgId: msgId, sender: sender);

  /// 文本内容
  final String text;

  @override
  String toString() {
    return "msgId: $msgId, sender: $sender, text: $text";
  }

  @override
  String getData() {
    return text;
  }
}

/// 消息数据包
class MsgPackData<T> extends MsgBase<T> {
  const MsgPackData({this.data, int msgId, sender})
      : super(msgId: msgId, sender: sender);

  final T data;

  @override
  T getData() {
    return data;
  }

  @override
  String toString() {
    return "msgId: $msgId, sender: $sender, data: $data";
  }
}

/// 消息处理器
class MsgHandlerBase implements MsgHandler {
  /// 已注册的消息ID列表
  List<int> _msgIds;

  List<int> get msgIds => _msgIds;

  /// 所有者
  final owner;

  /// 消息处理回调
  final MsgProcHandler onProcessMsg;

  MsgHandlerBase({this.owner, List<int> msgIds, this.onProcessMsg}) {
    this._msgIds = msgIds;
    if (this._msgIds == null) this._msgIds = new List<int>();
  }

  /// 处理消息
  void processMsg(MsgPack msg) async {
    if (onProcessMsg != null) onProcessMsg.processMsg(msg);
  }

  /// 异步发送消息
  void sendMsg(MsgPack msg) {
    if (msg == null) return;
    MsgEngine.instance.sendMsg(msg);
  }

  /// 同步发送消息
  void sendMsgSync(MsgPack msg) {
    if (msg == null) return;
    MsgEngine.instance.sendMsgSync(msg);
  }

  /// 释放资源
  void dispose() {
    if (_msgIds != null && (MsgEngine._instance != null)) {
      MsgEngine._instance.unRegisterHandler(this);
      _msgIds.clear();
    }
  }

  /// 添加消息ID
  void addMsgId(int msgId) {
    if (_msgIds.contains(msgId)) return;
    _msgIds.add(msgId);
    MsgEngine.instance.register(this, msgId);
  }

  /// 添加消息id列表
  void addMsgIds(List<int> msgId) {
    if (msgId == null || msgId.length == 0) return;
    if (_msgIds.length == 0) {
      for (int i = 0; i < msgId.length; i++) {
        int id = msgId[i];
        _msgIds.add(id);
        MsgEngine.instance.register(this, id);
      }
    } else {
      for (int i = 0; i < msgId.length; i++) {
        int id = msgId[i];
        if (_msgIds.contains(id)) continue;
        _msgIds.add(id);
        MsgEngine.instance.register(this, id);
      }
    }
  }

  /// 移除一个消息id
  void removeMsgId(int msgId) {
    if (_msgIds.remove(msgId)) {
      MsgEngine.instance.unRegister(msgId);
    }
  }

  /// 清空消息id
  void clearMsgId() {
    MsgEngine.instance.unRegisterHandler(this);
    _msgIds.clear();
  }
}

/// 消息处理器 （回调方式）
class MsgHandlerEvent extends MsgHandlerBase {
  final Map<int, MsgProcHandlerCallBack> _eventList =
  new Map<int, MsgProcHandlerCallBack>();

  MsgHandlerEvent({owner, List<int> msgIds})
      : super(owner: owner, msgIds: msgIds);

  @override
  void processMsg(MsgPack msg) {
    MsgProcHandlerCallBack callback = _eventList[msg.msgId];
    if (callback != null) callback(msg);
  }
}

/// 日志输出事件
typedef MsgEngineLogEvent(MsgEngine engine, String text);

/// 消息引擎
class MsgEngine {
  factory MsgEngine() {
    return _getInstance();
  }

  MsgEngine._internal();

  /// 单例的消息引擎
  static MsgEngine get instance => _getInstance();

  static MsgEngine _instance;

  static MsgEngine _getInstance() {
    if (_instance == null) {
      _instance = new MsgEngine._internal();
    }
    return _instance;
  }

  final Map<int, List<MsgProcHandler>> _msgHandlerMap =
  new Map<int, List<MsgProcHandler>>();
  final Queue<MsgPack> _msgQueue = new Queue<MsgPack>();

  /// 是否暂停
  bool isPause = false;

  /// 是否正在释放
  bool isDestroying = false;

  /// 消息分发速率。此值越大，消息分发速度越快，越容易卡界面
  int speedRatio = 250;

  /// 引擎自带的消息处理器
  MsgHandlerEvent msgHandler;

  /// 日志输出事件
  MsgEngineLogEvent onLog;

  bool _messageQueueRuning = false;
  int _lastStartMsgQueue = 0;

  /// 消息队列是否为空
  bool get isEmpty => _msgQueue.isEmpty;

  /// 消息队列长度
  int get msgQueueLength => _msgQueue.length;

  /// 开始消息引擎
  void start() async {
    if (_messageQueueRuning == false ||
        currentTimeMillis() - _lastStartMsgQueue > 5000) {
      _messageQueueRuning = false;
      isPause = false;
      _startMessageQueue();
    }
  }

  /// 暂停消息引擎
  void pause([bool value = true]) {
    isPause = value;
  }

  /// 停止消息引擎
  void stop() {
    isDestroying = true;
  }

  void _startMessageQueue() {
    if (_messageQueueRuning || isDestroying) return;
    _messageQueueRuning = true;
    _lastStartMsgQueue = currentTimeMillis();
    _messageQueueProcess();
  }

  bool __messageQueueProcessing = false;

  // 处理消息队列
  _messageQueueProcess() async {
    if (__messageQueueProcessing) return;
    __messageQueueProcessing = true;
    try {
      int _ref = 0;
      while (true) {
        //print("msg process.");
        if (isPause || _msgQueue.isEmpty) {
          return;
        } else {
          _ref++;
          if (_ref > speedRatio) {
            if (!isDestroying) {
              await Future.delayed(Duration(milliseconds: 20));
              if (isDestroying || _msgQueue.isEmpty) {
                __messageQueueProcessing = false;
                return;
              }
              _ref = 0;
            }
          }

          MsgPack msg = _msgQueue.first;
          try {
            _msgQueue.removeFirst();
            List<MsgProcHandler> o = _msgHandlerMap[msg.msgId];
            if (o != null && o.length > 0) sendMsgToHandler(o, msg);
          } catch (e) {
            _log(e.toString());
          }
        }
      }
    } catch (e) {
      _log(e.toString());
    } finally {
      __messageQueueProcessing = false;
    }
  }

  static int currentTimeMillis() {
    return DateTime.now().millisecondsSinceEpoch;
  }

  void _log(String msg) {
    if (onLog != null) onLog(this, msg);
  }

  void sendMsgToHandler(List<MsgProcHandler> handler, MsgPack msg) async {
    for (int i = 0; i < handler.length; i++) {
      try {
        handler[i].processMsg(msg);
      } catch (e) {
        _log("sendMsg: ${e.toString()}");
      }
    }
  }

  /// 发送消息
  void sendMsg(MsgPack msg) async {
    _msgQueue.add(msg);
    if (_messageQueueRuning && !isDestroying)
      _messageQueueProcess();
  }

  /// 同步发送消息
  void sendMsgSync(MsgPack msg) {
    sendMsgToHandler(getHandlerList(msg.msgId), msg);
  }

  /// 获取指定MsgID的消息处理器链表
  List<MsgProcHandler> getHandlerList(int msgId) {
    return _msgHandlerMap[msgId];
  }

  bool existHandler(List<MsgProcHandler> handlers, MsgProcHandler value) {
    return handlers.contains(value);
  }

  /// 消息处理器注册
  void register(MsgProcHandler msgHandler, int msgId) {
    if (msgHandler == null || msgId == 0) return;
    print('register: $msgId');
    List<MsgProcHandler> items = getHandlerList(msgId);
    if (items == null) {
      items = new List<MsgProcHandler>();
      items.add(msgHandler);
      _msgHandlerMap[msgId] = items;
    } else {
      if (!existHandler(items, msgHandler)) {
        items.add(msgHandler);
      }
    }
  }

  /// 消息处理器注册
  void registers(MsgProcHandler msgHandler, List<int> msgIds) {
    if (msgHandler == null || msgIds == null || msgIds.length == 0) return;
    for (int i=0; i<msgIds.length; i++) {
      int msgId = msgIds[i];
      if (msgId == 0) continue;
      register(msgHandler, msgId);
    }
  }

  /// 反注册消息处理器
  void unRegisterHandler(MsgProcHandler msgHandler) {
    if (msgHandler == null) return;
    if (msgHandler is MsgHandler) {
      List<int> keys = msgHandler.msgIds;
      if (keys == null || keys.length == 0) return;
      for (int i = 0; i < keys.length; i++) {
        int id = keys[i];
        List<MsgHandler> items = _msgHandlerMap[id];
        if (items != null) items.remove(msgHandler);
      }
    } else {
      List<int> keys = _msgHandlerMap.keys.toList();
      for (int i=0; i< keys.length; i++) {
        List<MsgProcHandler> items = _msgHandlerMap[keys[i]];
        if (items == null) continue;
        items.remove(msgHandler);
      }
    }
  }

  /// 反注册消息ID
  void unRegister(int msgId) {
    print('unRegister: $msgId, id count: ${_msgHandlerMap.length}');
    _msgHandlerMap.remove(msgId);
  }

  /// 解除所有已经注册的消息处理器
  void clearRegister() {
    _msgHandlerMap.clear();
  }

  /// 注册消息
  void regMsg(int msgId, MsgProcHandlerCallBack msgProcess) {
    if (msgHandler == null) msgHandler = new MsgHandlerEvent(owner: this);
    msgHandler._eventList[msgId] = msgProcess;
    msgHandler.addMsgId(msgId);
  }

  /// 取消注册的消息ID
  void unRegMsg(int msgId) {
    if (msgHandler == null) return;
    msgHandler.removeMsgId(msgId);
    msgHandler._eventList.remove(msgId);
  }
}


android/src/main/java/com/yangyxd/fluttermsgengine/FlutterMsgEnginePlugin.java
package com.yangyxd.fluttermsgengine;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** FlutterMsgEnginePlugin */
public class FlutterMsgEnginePlugin implements MethodCallHandler {
  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "flutter_msg_engine");
    channel.setMethodCallHandler(new FlutterMsgEnginePlugin());
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    } else {
      result.notImplemented();
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:flutter_msg_engine/flutter_msg_engine.dart';

void main() => runApp(new MyApp());

int SMsgId = 999;
int SMsgRef = 0;

// Use MsgProcessHandler process message.
class MyApp extends StatefulWidget implements MsgProcHandler<String> {

  @override
  _MyAppState createState() {
    MsgEngine.instance.register(this, 150);
    return new _MyAppState();
  }

  @override
  void processMsg(MsgPack<String> msg) {
    print("MyApp: " + msg.data);
  }
}

class _MyAppState extends State<MyApp> {

  @override
  void initState() {
    super.initState();
    // message start.
    MsgEngine.instance.start();
    // setting message log callback.
    MsgEngine.instance.onLog = ((MsgEngine engine, String text) {
      print(text);
    });
  }

  @override
  void dispose() {
    super.dispose();
    // stop message engine.
    MsgEngine.instance.stop();
  }

  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      routes: <String, WidgetBuilder>{
        "/main": (BuildContext context) => new MyHomePage(),
      },
      home: new MyHomePage(title: "HomePage"),
    );
  }

  static _test() {
    MsgEngine.instance.regMsg(111, (MsgPack msg) {
      print(msg.toString());
    });
    MsgEngine.instance.regMsg(115, (MsgPack msg) {
      print("115: ${msg.toString()}");
    });

    MsgEngine.instance.regMsg(118, (MsgPack msg) {
      print("118: ${msg.data}");
    });

    print(MsgEngine().toString());
    MsgPack msg = const MsgBase(msgId: 123, sender: null);
    print("${msg.toString()}");

    MsgEngine.instance.sendMsgSync(const MsgTextBase(msgId: 111, text: "Hello MsgEngine."));
    MsgEngine.instance.sendMsg(const MsgTextBase(msgId: 112, text: "empty"));
    MsgEngine.instance.sendMsg(const MsgTextBase(msgId: 111, text: "Hello MsgEngine 2."));
    MsgEngine.instance.sendMsg(const MsgTextBase(msgId: 113, text: "..."));
    MsgEngine.instance.sendMsg(const MsgTextBase(msgId: 115, text: "Hello MsgEngine 3."));
    MsgEngine.instance.sendMsg(const MsgTextBase(msgId: 150, text: "message 150."));

    MsgEngine.instance.sendMsg(const MsgPackData<int>(msgId: 150, data: 150));
    MsgEngine.instance.sendMsg(const MsgPackData<int>(msgId: 118, data: 520));

    MsgEngine.instance.sendMsg(MsgPackData<String>(msgId: SMsgId - 1, data: "message id is $SMsgRef"));
    MsgEngine.instance.sendMsg(MsgPackData<String>(msgId: SMsgId, data: "message id is $SMsgRef"));

    SMsgRef++;
  }
}

class MyHomePage extends StatefulWidget {
  final String title;
  MyHomePage({Key key, this.title}) : super(key: key);

  @override
  createState() => new _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> implements MsgProcHandler<String> {
  int _msgid;
  String _msg;

  @override
  void initState() {
    super.initState();
    _msgid = SMsgId;
    SMsgId++;
    MsgEngine.instance.register(this, _msgid);
  }

  @override
  void dispose() {
    super.dispose();
    MsgEngine.instance.unRegister(_msgid);
  }

  @override
  void processMsg(MsgPack<String> msg) {
    print(msg.data);
    setState(() {
      _msg = msg.data;
    });
  }

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      appBar: new AppBar(title: new Text(widget.title ?? "New Page")),
      body: new Padding(padding: const EdgeInsets.all(8.0), child: new Wrap(
        direction: Axis.vertical,
        spacing: 8.0,
        children: <Widget>[
          new Text("msgId: $_msgid, Data: $_msg"),
          new FlatButton(onPressed: () {
            _MyAppState._test();
          }, color: Colors.blue, splashColor: Colors.yellow, child: new Text("Test MsgEngine")),
          new MaterialButton(onPressed: () {
            Navigator.pushNamed(context, "/main");
            new Timer(const Duration(milliseconds: 1000), () {
              MsgEngine.instance.sendMsg(MsgPackData<String>(msgId: _msgid, data: "start new page."));
            });
          }, color: Colors.grey, child: new Text("Open New Page"))
        ],
      )),
    );
  }
}